cpp/src/arrow/compute/kernels/scalar_if_else_test.cc

```diff
@@ -0,0 +1,266 @@
+
+  for (auto&& cond : cond_datums) {
+    for (auto&& left : left_datums) {
+      for (auto&& right : right_datums) {
+        Datum exp;
+        GenerateExpected<TypeParam>(cond, left, right, &exp);
+        CheckIfElseOutput(cond, left, right, exp);
+      }
+    }
+  }
+}
+
+TEST_F(TestIfElseKernel, IfElseBooleanGen) {
+  auto type = boolean();
+
+  std::vector<Datum> cond_datums{ArrayFromJSON(boolean(), "[true, true, true, false]"),
+                                 ArrayFromJSON(boolean(), "[true, true, null, false]"),
+                                 MakeScalar(boolean(), true).ValueOrDie(),
+                                 MakeNullScalar(boolean())};
+
+  std::vector<Datum> left_datums{ArrayFromJSON(type, "[false, false, false, false]"),
+                                 ArrayFromJSON(type, "[false, false, null, false]"),
+                                 MakeScalar(type, false).ValueOrDie(),
+                                 MakeNullScalar(type)};
+
+  std::vector<Datum> right_datums{ArrayFromJSON(type, "[true, true, true, true]"),
+                                  ArrayFromJSON(type, "[true, true, true, null]"),
+                                  MakeScalar(type, true).ValueOrDie(),
+                                  MakeNullScalar(type)};
+
+  for (auto&& cond : cond_datums) {
+    for (auto&& left : left_datums) {
+      for (auto&& right : right_datums) {
+        Datum exp;
+        GenerateExpected<BooleanType>(cond, left, right, &exp);
+        CheckIfElseOutput(cond, left, right, exp);
+      }
+    }
+  }
+}
+
+TYPED_TEST(TestIfElsePrimitive, IfElseBooleanRand) {
+  auto type = boolean();
+  random::RandomArrayGenerator rand(/*seed=*/0);
+  int64_t len = 1000;
+  auto cond = std::static_pointer_cast<BooleanArray>(
+      rand.ArrayOf(boolean(), len, /*null_probability=*/0.01));
+  auto left = std::static_pointer_cast<BooleanArray>(
+      rand.ArrayOf(type, len, /*null_probability=*/0.01));
+  auto right = std::static_pointer_cast<BooleanArray>(
+      rand.ArrayOf(type, len, /*null_probability=*/0.01));
+
+  BooleanBuilder builder;
+  for (int64_t i = 0; i < len; ++i) {
+    if (!cond->IsValid(i) || (cond->Value(i) && !left->IsValid(i)) ||
+        (!cond->Value(i) && !right->IsValid(i))) {
+      ASSERT_OK(builder.AppendNull());
+      continue;
+    }
+
+    if (cond->Value(i)) {
+      ASSERT_OK(builder.Append(left->Value(i)));
+    } else {
+      ASSERT_OK(builder.Append(right->Value(i)));
+    }
+  }
+  ASSERT_OK_AND_ASSIGN(auto expected_data, builder.Finish());
+
+  CheckIfElseOutput(cond, left, right, expected_data);
+}
+
+TEST_F(TestIfElseKernel, IfElseNull) {
+  CheckIfElseOutput(ArrayFromJSON(boolean(), "[null, null, null, null]"),
+                    ArrayFromJSON(null(), "[null, null, null, null]"),
+                    ArrayFromJSON(null(), "[null, null, null, null]"),
+                    ArrayFromJSON(null(), "[null, null, null, null]"));
+}
+
+TEST_F(TestIfElseKernel, IfElseWithOffset) {
```

Comment:
You could also do this in the random tests by testing in a loop, say, every offset from 0 to 8 at the end.